tests/template_tests/test_engine.py

```diff
@@ -10,6 +10,36 @@
 OTHER_DIR = os.path.join(ROOT, 'other_templates')
 
 
+class EngineReprTest(SimpleTestCase):
+
+    def test_repr(self):
+        app_dirs = True
+        loaders = [
+            'django.template.loaders.filesystem.Loader',
+            'django.template.loaders.app_directories.Loader',  # since app_dir is True, this is required
+        ]
+        context_processors = []
+        engine = Engine(
+            dirs=[TEMPLATE_DIR],
+            app_dirs=app_dirs,
+            loaders=None,
+            context_processors=context_processors,
+            debug=True,
+        )
+
+        msg = (
+            '<%(class)s dirs=%(dirs)s, app_dirs=%(app_dirs)s, loaders=%(loaders)s,'
+            ' context_processors=%(context_processors)s>'
+        ) % {
+            'class': 'Engine',
+            'app_dirs': app_dirs,
+            'loaders': loaders,
+            'context_processors': context_processors,
+            'dirs': [TEMPLATE_DIR],
+        }
+        self.assertEqual(repr(engine), msg)
```

Comment:
I would assert with an explicit representation (except `TEMPLATE_DIR` which depends on OS) e.g.
```python
        self.assertEqual(repr(engine), (
            f"<Engine dirs=[{TEMPLATE_DIR!r}] app_dirs=True "
            f"context_processors=[] loaders=["
            f"'django.template.loaders.filesystem.Loader' "
            f"'django.template.loaders.app_directories.Loader'] "
            ...
        ))
```

Reply:
I have done what you asked for, but in my opinion this sort of long string with double quotes and single quotes mixed in between doesn't look too readable to me.

Reply:
Also, in this sort of assertions may be potentially hard to pin-point the accurate diff between long strings if the test fails. as it happened to me, I actually had to interchange the single quotes and double quotes for this test to pass i.e. when single quotes were put around the edges and double quotes for interpolation, the tests actually failed. I'm not really sure why this happened, but seemed quite weird to me.

Reply:
Just to reproduce what I said above, you can test this script:
```python
# t.py
import unittest

class LongStringTest(unittest.TestCase):
    a_list = ['a', 'b']
    a_tuple = ('a', 'b')
    a_complex = [('a', ['b', 'c'])]
    some_long_string = (
        f'List {a_list!r}'
        f'Tuple {a_tuple!r}'
        f'Complex {a_complex!r}'
    )
    def test_single_quotes_at_ends_with_double_quotes_for_interpolation(self):
        """This test fails"""
        self.assertEqual(
            self.some_long_string,
            (
                'List ["a", "b"]'
                'Tuple ("a", "b")'
                'Complex [("a", ["b", "c"])]'
            )
        )

    def test_double_quotes_at_ends_with_single_quotes_for_interpolation(self):
        """This test passes"""
        self.assertEqual(
            self.some_long_string,
            (
                "List ['a', 'b']"
                "Tuple ('a', 'b')"
                "Complex [('a', ['b', 'c'])]"
            )
        )
```
The output on my machine:
```sh
> python -m unittest t 
.F
======================================================================
FAIL: test_single_quotes_at_ends_with_double_quotes_for_interpolation (t.LongStringTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/django/tests/t.py", line 13, in test_single_quotes_at_ends_with_double_quotes_for_interpolation
    self.assertEqual(
AssertionError: "List ['a', 'b']Tuple ('a', 'b')Complex [('a', ['b', 'c'])]" != 'List ["a", "b"]Tuple ("a", "b")Complex [("a", ["b", "c"])]'
- List ['a', 'b']Tuple ('a', 'b')Complex [('a', ['b', 'c'])]
?       ^ ^  ^ ^        ^ ^  ^ ^           ^ ^   ^ ^  ^ ^
+ List ["a", "b"]Tuple ("a", "b")Complex [("a", ["b", "c"])]
?       ^ ^  ^ ^        ^ ^  ^ ^           ^ ^   ^ ^  ^ ^


----------------------------------------------------------------------
Ran 2 tests in 0.001s

FAILED (failures=1)

```

I think this is probably a bug in python maybe, will try to investigate further.